src/sqlfluff/dialects/dialect_clickhouse.py

```diff
@@ -46,3 +46,31 @@ class CTEDefinitionSegment(ansi.CTEDefinitionSegment):
             Ref("SingleIdentifierGrammar"),
         ),
     )
+
+
+class FromExpressionElementSegment(ansi.FromExpressionElementSegment):
+    """A table expression.
+
+    Overridden from ANSI to allow FINAL modifier.
+    https://clickhouse.com/docs/en/sql-reference/statements/select/from#final-modifier
+    """
+
+    type = "from_expression_element"
+    match_grammar: Matchable = Sequence(
+        Ref("PreTableFunctionKeywordsGrammar", optional=True),
+        OptionallyBracketed(Ref("TableExpressionSegment")),
+        Ref(
+            "AliasExpressionSegment",
+            exclude=OneOf(
+                Ref("SamplingExpressionSegment"),
+                Ref("JoinLikeClauseGrammar"),
+                Ref.keyword("Final"),
+            ),
+            optional=True,
+        ),
+        Ref.keyword("Final", optional=True),
+        # https://cloud.google.com/bigquery/docs/reference/standard-sql/arrays#flattening_arrays
+        Sequence("WITH", "OFFSET", Ref("AliasExpressionSegment"), optional=True),
```

Comment:
To avoid repeating the whole grammar we can do things like this (untested):

```py
    match_grammar = ansi.FromExpressionElementSegment.match_grammar.copy(
        insert=[
            Ref.keyword("Final", optional=True),
        ],
        before=Sequence("WITH", "OFFSET", Ref("AliasExpressionSegment"), optional=True),,
    )
```

If that doesn't work then happy to merge as is.

Reply:
What about the line 67 ? 
I found that i had to repeat it to keep the support for alias and final. 
Still new to the parsing grammar, happy to find better solutions :)

Reply:
What about the line 67 ?
I found that i had to repeat it to keep the support for alias and final.
Still new to the parsing grammar, happy to find better solutions :)

Reply:
This would replace lines 59-76 if it works.

Reply:
Unfortunately without the extra  `Ref.keyword("Final"),` L67 in the `Ref("AliasExpressionSegment")` FINAL is detected as an `alias_expression: identifier: final`

Reply:
Ah got you. Missed that. OK let's merge as is!